DigitalLearningSolutions.Web/Controllers/TrackingSystem/CourseSetup/CourseSetupController.cs

```diff
@@ -94,5 +113,363 @@ public IActionResult AllCourseStatistics()
+                model.SelectAllSections();
+                var availableSections = GetSectionModelsWithAllContentEnabled(model, data.Application.DiagAssess);
+                data!.SetSectionContentModels = availableSections.ToList();
+            }
+
+            if (!ModelState.IsValid)
+            {
+                return View("AddNewCentreCourse/SetCourseContent", model);
+            }
+
+            data!.SetCourseContentModel = model;
+            TempData.Set(data);
+
+            return RedirectToAction(model.IncludeAllSections ? "Summary" : "SetSectionContent");
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<AddNewCentreCourseData>))]
+        [HttpGet]
+        [Route("AddCourse/SetSectionContent")]
+        public IActionResult SetSectionContent(int sectionIndex)
+        {
+            var data = TempData.Peek<AddNewCentreCourseData>();
+
+            var section = data!.SetCourseContentModel!.GetSelectedSections().ElementAt(sectionIndex);
+            var tutorials = tutorialService.GetTutorialsForSection(section.Id).ToList();
+            var showDiagnostic = data.Application!.DiagAssess;
+            var model = new SetSectionContentViewModel(section, sectionIndex, showDiagnostic, tutorials);
+
+            return !tutorials.Any()
+                ? RedirectToNextSectionOrSummary(model.Index, data.SetCourseContentModel)
+                : View("AddNewCentreCourse/SetSectionContent", model);
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<AddNewCentreCourseData>))]
+        [HttpPost]
+        [Route("AddCourse/SetSectionContent")]
+        public IActionResult SetSectionContent(
+            SetSectionContentViewModel model,
+            string action
+        )
+        {
+            if (action == SaveAction)
+            {
+                return SaveSectionAndRedirect(model);
+            }
+
+            var bulkSelectResult = EditCourseSectionHelper.ProcessBulkSelect(model, action);
+            return bulkSelectResult ?? View("AddNewCentreCourse/SetSectionContent", model);
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<AddNewCentreCourseData>))]
+        [HttpGet]
+        [Route("AddCourse/Summary")]
+        public IActionResult Summary()
+        {
+            var data = TempData.Peek<AddNewCentreCourseData>();
+
+            var model = new SummaryViewModel(data!);
+
+            return View("AddNewCentreCourse/Summary", model);
+        }
+
+        [ServiceFilter(typeof(RedirectEmptySessionData<AddNewCentreCourseData>))]
+        [HttpPost]
+        [Route("AddCourse/Summary")]
+        public IActionResult? Summary(SummaryViewModel model)
+        {
+            var data = TempData.Peek<AddNewCentreCourseData>()!;
+
+            using var transaction = new TransactionScope();
+
+            var centreId = User.GetCentreId();
+            var customisationId = courseService.CreateNewCentreCourse(

```

Comment:
Theres a fair amount of stuff here that could be done in the service. Additionally, seeing as the bulk comes from `data` (a AddNewCentreCourseData), could we just pass `data` in as a single parameter and leave the service to do all of this parsing.

Reply:
The issue with that is that AddNewCentreCourseData and the models it contains all exist in the web project rather than the data project. I've pulled some of this out into a private method to make it cleaner, does that look okay?

Reply:
Ok, I see. The extra changes into a private method help, thanks!